src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -1575,8 +1616,109 @@ public Subscriptions(IDisposable subscription, Subscriptions? next)
 #endregion
 
         private FilterAndTransform? _specs;                 // Transformation specifications that indicate which sources/events are forwarded.
-        private FilterAndTransform? _activitySourceSpecs;   // ActivitySource Transformation specifications that indicate which sources/events are forwarded.
+        private FilterAndTransform? _rawActivitySourceSpecs;
+        private Dictionary<string, ActivitySourceFilterAndTransformSpecs>? _activitySourceSpecsBySourceName;
```

Comment:
@tarekgh Could we use `FrozenDictionary` here? Thinking it would make the sampling logic execute faster when creating spans using out-of-proc sampling. There is a second `Dictionary` lookup when specific activity names are also defined (ex `[AS]MySource+MyActivity/-ParentRatioSampler(0.05)`)

Reply:
The only concern is the new dependencies. We added Linq dependency before and we had to revert it. If @noahfalk is ok with such dependency. 

CC @stephentoub in case he has more feedback using the frozen collections.

Reply:
With regards to viability, this dictionary is created once, used for the remainder of the process, and lookup speed is critical?

Reply:
I'd be hesitant to add that dependency. At the very least before doing that I'd want to understand what the current performance looks like, what is the goal and where does profiling say the current hot-spots are. 

My best guess is that if you want to make this faster we've probably optimizations that are as good or better and don't need a new dependency. For example instead of using a two-tiered Dictionary you could do a single-tier Dictionary that maps (Source,ActivityName) -> Spec. When you encounter a (Source,ActivityName) tuple for the first time you'd scan the entire set of specs, pick the one which matches best and cache it. Every future time it will be in the dictionary.